polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -126,208 +173,163 @@ class MemorySelector {
-
-  /// Determine whether the memref \access exhibits temporal reuse.
-  bool hasTemporalReuse(const affine::MemRefAccess &access,
+  /// Determine whether \p memRefAccess exhibits temporal reuse.
+  bool hasTemporalReuse(const affine::MemRefAccess &memRefAccess,
                         const SmallVectorImpl<Value> &threadVars) const;
 
 private:
-  const MemoryAccessAnalysis &memAccessAnalysis;
-
+  MemoryAccessAnalysis &memAccessAnalysis;
+  AliasAnalysis &aliasAnalysis;
   DataFlowSolver &solver;
 
-  /// Collects all memory accesses for a given memref value.
-  DenseMap<Value, SmallVector<affine::MemRefAccess>> accesses;
-
   /// The preferred memory space for each memref access.
-  DenseMap<const Operation *, MemorySpace> accessToMemSpace;
+  DenseMap<Value, MemorySpace> memRefAccessToMemSpace;
 };
 
 std::optional<MemorySelector::MemorySpace>
-MemorySelector::selectMemorySpace(Value memref) const {
+MemorySelector::getMemorySpace(Value memref) const {
   assert(isa<MemRefType>(memref.getType()) && "Expecting a memref");
 
-  auto it = accesses.find(memref);
-  if (it == accesses.end())
+  auto it = memRefAccessToMemSpace.find(memref);
+  if (it == memRefAccessToMemSpace.end())
     return std::nullopt;
-
-  auto numShared = [this](ArrayRef<affine::MemRefAccess> accesses) {
-    return llvm::count_if(accesses, [this](affine::MemRefAccess access) {
-      auto it = accessToMemSpace.find(access.opInst);
-      if (it == accessToMemSpace.end())
-        return false;
-      return (it->second == MemorySpace::Shared);
-    });
-  };
-
-  /// Recommend shared memory if at least half of the accesses for this memref
-  /// should use shared memory.
-  ArrayRef<affine::MemRefAccess> accesses = it->second;
-  if (numShared(accesses) >= std::ceil((double)accesses.size() / 2))
-    return MemorySpace::Shared;
-
-  return MemorySpace::Global;
+  return it->second;
 }
 
 void MemorySelector::analyze(LoopLikeOpInterface loop, AccessKind accessKind) {
-  assert(accesses.empty() && accessToMemSpace.empty() &&
-         "Expecting empty maps");
-
   // Collect the global thread ids used in the function the loop is in.
   auto funcOp = loop->template getParentOfType<FunctionOpInterface>();
   SmallVector<Value> threadVars =
       memAccessAnalysis.getThreadVector(funcOp, solver);
 
   // Collect candidate memref accesses in the loop.
+  DenseMap<Value, SmallVector<affine::MemRefAccess>> memRefToMemRefAccesses;
   loop->walk<WalkOrder::PreOrder>([&](Operation *op) {
     if (!isa<affine::AffineLoadOp, affine::AffineStoreOp>(op))
       return;
 
-    affine::MemRefAccess access(op);
-    addMemRefAccess(access);
+    affine::MemRefAccess memRefAccess(op);
+
+    auto it = memRefToMemRefAccesses.find(memRefAccess.memref);
+    if (it == memRefToMemRefAccesses.end())
+      memRefToMemRefAccesses[memRefAccess.memref] = {memRefAccess};
+    else
+      memRefToMemRefAccesses[memRefAccess.memref].push_back(memRefAccess);
```

Comment:
```suggestion
      memRefToMemRefAccesses[memRefAccess.memref].push_back(memRefAccess);
```
can you check if this work?

Reply:
done.